### ISSUE TITLE
valid-ngdoc

### ISSUE THREAD

**Issue by szgabsz91** [Gillespie59/eslint-plugin-angular#335]
Hi,

I have an ESLint configuration for an AngularJS project with the valid-jsdoc rule switched on.

Let's assume I have an ngdoc comment like this:

    /**
     * @ngdoc service
     * @name MyService
     * @requires $http
     */
    function MyService($http) {
    }
    angular.module('mymodule').factory('MyService', MyService);

With the valid-jsdoc rule, I get the following error:

    Missing JSDoc for parameter '$http'

If I add an extra unnecessary annotation, the error disappears:

    @param {object} [$http] AngularJS $http service

This made make think if it was a good idea to have a custom valid-ngdoc rule in eslint-plugin-angular that would use the original one somehow, but it would know that in this case if there's a @requires then the @param is unnecessary.

There might be other issues as well, but this is what I've found so far.

I didn't investigate a lot on how this would be achievable, but I'm curious if there are other thoughts/solutions.

Thanks!

**Comment by remcohaszing** [comment]
I have been thinking about implementing this. For me this led to the final though that this would be a pointless feature.

The only reason to lint this, is because Dgeni lacks the ability to extract those tags from the AST of AngularJS code.

`eslint` parses the JS code using `espree` and then traverses the AST using `estraverse`. Creating an `eslint` rule to parse these tags would have about the same difficulty level as creating a Dgeni package.

I would definitely find it useful to lint this and would welcome a PR implementing this, but I'd find it much more useful if someone would make a PR to [dgeni-packages](https://github.com/angular/dgeni-packages) so those tags won't be necessery anymore at all.

**Comment by szgabsz91** [comment]
I'm not sure I totally understand, I'm just trying to explore the connections among these packages.

So eslint's valid-jsdoc rule is responsible for linting the documentation comments. However, eslint cannot have framework specific rules, that's why I though this package would be a great candidate.

And I might be wrong, but it seems to me that eslint parses the documentation comments with doctrine (and not dgeni) which is also an eslint package that does not depend on dgeni-packages. The parsing takes place [here](https://github.com/eslint/eslint/blob/master/lib/rules/valid-jsdoc.js#L192) and for the above example, the `jsdoc` variable will be:

    {
        description: '',
        tags: [
            { title: 'ngdoc', description: 'service' },
            { title: 'name', description: null, name: 'MyService' },
            { title: 'requires', description: null, name: '$http' }
        ]
    }

I managed to debug this, that's why I thought an "extended" rule here would be OK because the modified logic could check if `@requires` is there for `$http`, so it wouldn't throw an error because of the lack of `@param`.

Of course the proper implementation would be crucial so that no code duplication is present here.

But after your comment, I think I don't understand what's the connection of eslint/eslint-plugin-angular/espree/estraverse to dgeni-pacakges and how would this possible PR you mentioned look like? Something like replacing `@param` with `@requires` in certain cases, e.g. for services, controllers, etc. and keep it for methods? And even if dgeni-packages would accept this PR, how would the eslint ecosystem benefit from it? By replacing doctrine with dgeni?

Thanks for your response, I'm really curious. :)

**Comment by remcohaszing** [comment]
Now that I re-read my previous reply, I admit it was a pretty confusing answer.

szgabsz91:
> But after your comment, I think I don't understand what's the connection of eslint/eslint-plugin-angular/espree/estraverse to dgeni-pacakges and how would this possible PR you mentioned look like?

* There is no connection between ESLint and Dgeni.
* Espree / estraverse is a library that just happens to be used internally by both. It's actually not that important, but it shows that creating an ESLint rule is pretty similar to implementing an tool to extract the information from the AST instead of comments.


***So what is Dgeni?***

Dgeni is the successor of `ngdocs`. Dgeni is maintained by the AngularJS team and is used to extract `@ngdoc` tags from both source code and `.ngdoc` files and render this to something useful, e.g. and HTML file. This is used by several projects, such as Angular Material, Protractor and AngularJS itself.

An simple example of such ngdoc would be:

```js
/**
 * @ngdoc service
 * @name MyService
 * @module mymodule
 * @requires $http
 *
 * @description
 * It's all mine!
 */
function MyService($http) {
  return {
    myMethod: myMethod
  };

  /**
   * @ngdoc method
   * @name MyService#myMethod
   *
   * @description
   * Make sure it's mine.
   *
   * @returns {string} `MINE!`
   */
  function myMethod() {
    return 'MINE!';
  }
}

angular.module('mymodule').factory('MyService', MyService);
```

I think this is a lot of boilerplate, or actually code duplication.


***What would such an ESLint rule do?***

An ESLint rule could detect missing / extraneous / invalid tags (assuming somewhat default configurations are used).

```js
/**
 * @ngdoc service
 * @name MyService
 * @module yourmodule                  // <-- Bad module name
 * @requires $http
 * @requires $q                        // <-- Extraneous requires
 *
 * @description
 * It's all mine!
 */
function MyService($http, $resource) { // <-- Forgot to document $resource
  return {
    yourMethod: myMethod
  };

  /**
   * @ngdoc method
   * @name MyService#myMethod          // <-- Bad method name
   *
   * @description
   * Make sure it's mine.
   *
   * @returns {string} `MINE!`
   */
  function myMethod() {
    return 'MINE!';
  }
}

angular.module('mymodule').factory('MyService', MyService);
```

This is basically what you're suggesting. I totally agree this would be a great feature. I personally spend a lot of time to check if docs are being updated correctly in code reviews. I would **love** to have a rule to detect this for me.


***Why I doubt this is useful***

If a tool exists which can validate the correctness of the documentation against the code, then it should also be possible to create a tool which can extract the needed data from the code. This means developers don't even have to keep track of the documentation manually.

Instead, the code would look like this:

```js
/**
 * @description
 * It's all mine!
 */
function MyService($http, $resource) {
  return {
    yourMethod: myMethod
  };

  /**
   * @description
   * Make sure it's mine.
   *
   * @returns {string} `MINE!`
   */
  function myMethod() {
    return 'MINE!';
  }
}

angular.module('mymodule').factory('MyService', MyService);
```

Actually this already exists partially in the [`angular-docs`](https://github.com/angular/dgeni-packages/tree/angular-docs) branch of `dgeni-packages`. It just hasn't been finished / touched for months. So even though I would like such a rule and would definitely use it myself, I think contributing there would be far more useful.

**Comment by szgabsz91** [comment]
Aha, it's far more clear now! :+1: I just didn't know about this new dgeni proposal.

I totally understand, actually a couple of years ago it was a bit weird for me that in this system the documentation annotations and the code itself are totally separate things. On the other hand I accepted it because it is a bit more difficult to extract information from JS code than from Java or C# for instance.

So yes, I get it now, if there's a work in progress there, it would be cool to see the outcome, because this would definitely make things much easier.

I guess I'll deep dive into the code of the dgeni-packages branch you suggest in my free time, and contribute if I can. So if you think, this one can be closed.

And thanks again for the detailed info!

**Comment by remcohaszing** [comment]
@szgabsz91 This turned out to be easier than I originally imagined. I've created a basic implementation already in the referencing PR above. I think it'll be available soon, if I have enough free time. :wink:
